### DISCUSSION TITLE
Installing extensions through NPM

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3108] [Feature Requests]
This needs research...

**Comment by nickrum** [comment]
The plan currently is to allow npm packages with a name of `directus-extension-*` or `@directus/extension-*` to be installed into the npm package where directus has been installed. Those packages must have an additional `directus:extension` field inside their `package.json` file.
This field should have the following structure:

```json
"directus:extension": {
    "type": "interface",
    "path": "dist/index.js",
    "host": "^9.0.0"
}
```

or

```json
"directus:extension": {
    "type": "pack",
    "host": "^9.0.0"
}
```

`type` specifies the type of the extension. It can be one of "interface", "display", "layout", "module", "endpoint", "hook" or "pack".
`path` specifies the entrypoint for the extension. For interface, display, layout or module extensions this must be a single bundled file.
`host` specifies the directus version with which the extension is compatible as a semver string. The compatibility will be checked at load time of the extension.

If the specified extension type is "pack", directus will load any npm packages with a name of `directus-extension-*` or `@directus/extension-*` listed in the package's dependencies as extensions. Those extensions itself can again be of type "pack".
Directus will resolve nested extensions using the parent package as the root of the resolve operation. If an extension was found multiple times while resolving the dependency tree, loading extensions will fail.

(cc @rijkvanzanten, @eliashussary)

**Reply by benhaynes** [reply]
A few non-technical questions:
* Do we feel that this `directus-extension-*` naming convention is too lengthy? Just wondering if we should consider other options, like `directus-ext-*`. Maybe it's just me, but this seems so long: 
    *  `@benhaynes/directus-extension-seating-chart`
* Do we prefer `pack` or `bundle`?
* With that SemVer convention, can you go in both ways? (eg: _after X but also before Y_) Or is the expectation that extensions should always be kept up to date with Directus latest?

**Reply by rijkvanzanten** [reply]
> Do we feel that this directus-extension-* naming convention is too lengthy? Just wondering if we should consider other options, like directus-ext-*. Maybe it's just me, but this seems so long:
@benhaynes/directus-extension-seating-chart

Yes we can, but generally speaking there's nothing inherently wrong with a slightly longer name. I'd prefer clarity over brevity.

> Do we prefer pack or bundle?

`bundle` 👍🏻 

> With that SemVer convention, can you go in both ways? (eg: after X but also before Y) Or is the expectation that extensions should always be kept up to date with Directus latest?

Yes to both! You can use `>=9.0.0 <10.0.0` to signify a range, but we should definitely aim to make extensions easy enough to support where they can ideally be updated and supported across as many versions as possible

**Reply by nickrum** [reply]
One thing I haven't thought of yet: Should we actually support scoped packages like `@benhaynes/directus-extension-seating-chart`?

> Do we prefer `pack` or `bundle`?

To me "bundle" sounds more like a npm package _containing_ multiple extensions (which was the initial idea for npm extensions), instead of an _empty_ package acting as a _virtual_ container. Also, this way we could potentially support the former without having to come up with a new name for it. I took some inspiration from `vscode`, where such plugin types are called "packs".

> Yes to both! You can use >=9.0.0 <10.0.0 to signify a range, but we should definitely aim to make extensions easy enough to support where they can ideally be updated and supported across as many versions as possible

Maybe we should have a separate "extension api version" that tracks added features and breaking changes for the extension api only?

**Reply by rijkvanzanten** [reply]
> One thing I haven't thought of yet: Should we actually support scoped packages like @benhaynes/directus-extension-seating-chart?

Yeah I think so. I always prefer to publish my own packages under `@rijk/whatever` 

> To me "bundle" sounds more like a npm package containing multiple extensions (which was the initial idea for npm extensions), instead of an empty package acting as a virtual container. Also, this way we could potentially support the former without having to come up with a new name for it. I took some inspiration from vscode, where such plugin types are called "packs".

Oh I didn't fully grasp that that top level extension would be "empty" with nothing more than pointers to the actual extensions. I was personally thinking that you'd use a single extension as the "entrypoint" for use, which in turn loads in a bunch of other extensions (a module would be the perfect "entrypoint" which in turn has a dependency on x extensions and y hooks)

> Maybe we should have a separate "extension api version" that tracks added features and breaking changes for the extension api only?

I get where you're coming from, but that feels a little confusing to me. I'd rather keep it tightly coupled to the actual Directus version, as that's easier to reason about (same reason why all packages under the Directus umbrella share version, instead of having API version 3.1.61, app 5.4.51, and SDK 2.1.957)

**Reply by nickrum** [reply]
> Yeah I think so. I always prefer to publish my own packages under `@rijk/whatever`

Alright 👍

> Oh I didn't fully grasp that that top level extension would be "empty" with nothing more than pointers to the actual extensions. I was personally thinking that you'd use a single extension as the "entrypoint" for use, which in turn loads in a bunch of other extensions (a module would be the perfect "entrypoint" which in turn has a dependency on x extensions and y hooks)

The problem is that we do not have a traditional dependency tree where each sub-dependency is only imported by it's parent. It's more like a tree of peer-dependencies which is less confusing and easier to handle by using virtual container packages and only allowing for actual extension packages in the leaves of such a tree. We could support your use case by allowing to specify peer-dependencies in the `package.json` of module/endpoint/etc. extensions.

> I get where you're coming from, but that feels a little confusing to me. I'd rather keep it tightly coupled to the actual Directus version, as that's easier to reason about (same reason why all packages under the Directus umbrella share version, instead of having API version 3.1.61, app 5.4.51, and SDK 2.1.957)

Thinking more about it, I came to the same conclusion!

**Comment by rijkvanzanten** [comment]
This has since been implemented!
